File: 1-arrays&strings/exercises/is_first_come_first_served.py
def is_first_come_first_served(take_out_orders, dine_in_orders, served_orders):
    take_out_orders_index = 0
    dine_in_orders_index = 0
    take_out_orders_max_index = len(take_out_orders) - 1
    dine_in_orders_max_index = len(dine_in_orders) - 1

    for order in served_orders:
        # If we still have orders in take_out_orders
        # and the current order in take_out_orders is the same
        # as the current order in served_orders
        if take_out_orders_index <= take_out_orders_max_index and order == take_out_orders[take_out_orders_index]:
            take_out_orders_index += 1

        # If we still have orders in dine_in_orders
        # and the current order in dine_in_orders is the same
        # as the current order in served_orders
        elif dine_in_orders_index <= dine_in_orders_max_index and order == dine_in_orders[dine_in_orders_index]:
            dine_in_orders_index += 1

        # If the current order in served_orders doesn't match the current
        # order in take_out_orders or dine_in_orders, then we're not serving first-come,
        # first-served.
        else:
            return False

    # Check for any extra orders at the end of take_out_orders or dine_in_orders
    if dine_in_orders_index != len(dine_in_orders) or take_out_orders_index != len(take_out_orders):
        return False

    # All orders in served_orders have been "accounted for"
    # so we're serving first-come, first-served!
    return True

if __name__ == '__main__':
    print(is_first_come_first_served([17, 8, 24], [12, 19, 2], [17, 8, 12, 19, 24, 2]))
    print(is_first_come_first_served([1, 3, 5], [2, 4, 6], [1, 2, 4, 6, 5, 3]))

File: 1-arrays&strings/exercises/is_first_come_first_served.go
package main

func isFirstComeFirstServed(takeOut []int, dineIn []int, served []int) bool {

	indexTakeOut := 0
	indexDineIn := 0
	endTakeOut := len(takeOut)
	endDineIn := len(dineIn)

	for _, order := range served {
		if (indexTakeOut < endTakeOut) && (order == takeOut[indexTakeOut]) {
			indexTakeOut += 1
		} else if (indexDineIn < endDineIn) && (order == dineIn[indexDineIn]) {
			indexDineIn += 1
		} else {
			return false
		}
	}
	if (indexTakeOut != endTakeOut) && (indexDineIn != endDineIn) {
		return false
	}
	return true
}
